include/deal.II/meshworker/scratch_data.h

```diff
@@ -0,0 +1,1301 @@
+
+    /**
+     * Find out if we store an object with given name.
+     */
+    bool
+    stores_object_with_name(const std::string &name) const;
+
+    /**
+     * Return the quadrature points of the internal FEValues object.
+     */
+    const std::vector<Point<spacedim>> &
+    get_quadrature_points() const;
+
+    /**
+     * Return the JxW values of the internal FEValues object.
+     */
+    const std::vector<double> &
+    get_JxW_values() const;
+
+    /**
+     * Return the JxW values of the neighbor FEValues object.
+     */
+    const std::vector<double> &
+    get_neighbor_JxW_values() const;
+
+    /**
+     * Return the last computed normal vectors.
+     */
+    const std::vector<Tensor<1, spacedim>> &
+    get_normal_vectors() const;
+
+    /**
+     * Return the last computed normal vectors on the neighbor.
+     */
+    const std::vector<Tensor<1, spacedim>> &
+    get_neighbor_normal_vectors();
+
+    /**
+     * Print the contents of the internal cache.
+     */
+    template <class Stream>
+    void
+    print_info(Stream &os);
+
+    /**
+     * Extract the local dof values associated with the internally initialized
+     * cell.
+     *
+     * Before you call this function, you have to make sure you have previously
+     * called one of the reinit() functions.
+     *
+     * At every call of this function, a new vector of dof values is generated
+     * and stored internally, unless a previous vector with the same name is
+     * found. If this is the case, the content of that vector is overwritten.
+     *
+     * If you give a unique @p global_vector_name, then for each cell you are
+     * guaranteed to get a unique vector of independent dofs of the same type
+     * as the dummy variable. If you use an automatic differentiation number
+     * type (like Sacado::Fad::DFad<double>,
+     * Sacado::Fad::DFad<Sacado::Fad::DFad<double>>, etc.) this method will
+     * also initialize the independent variables internally, allowing you
+     * to perform automatic differentiation.
+     *
+     * Once you have extracted the local dof values, you can later access the
+     * same dof values, without reinitializing them, by calling the method
+     * get_local_dof_values() with the same @p global_vector_name argument
+     * you passed here.
+     *
+     * Notice that using this initialization strategy renders the use of this
+     * ScratchData object incompatible with the AD helper classes (since they
+     * do their own data management). In particular, it is necessary for the
+     * user to manage all of the AD data themselves (both before and after this
+     * call).
+     */
+    template <typename VectorType, typename Number = double>
+    const std::vector<Number> &
+    extract_local_dof_values(const std::string &global_vector_name,
+                             const VectorType & input_vector,
+                             const Number       dummy = Number(0));
```

Comment:
C++11 allows you to have this template argument without having a corresponding function argument. Since I don't think anyone will ever want to pass anything other than `Number(0)` in this place, you might as well remove the argument. If a user wants to have a different type for `Number`, they will then have to explicitly provide the template argument.

The current default of the template argument is `double`. Why not use `typename VectorType::value_type`?

Reply:
As an aside, I think this interface is a bit clunky. What do you think about not returning a vector from this function? Then one would have to call `extract_...` once in the beginning, which would do the extraction. Every use of the extracted data would then require a call to `get_...`. This way, you'd call the same function every time, rather than calling `extract_...` the first time and `get_...` later on.

I'm thinking about this from the perspective of refactoring code previously written, where one often moves code blocks around. This becomes awkward/wrong if one has to call different functions at different places in the same code block.

Reply:
The reason for providing the argument to the function as well was to avoid the cumbersome:
~~~
auto &v = scratch.template get_values<Number>("name", sol);
~~~
and allow simply
~~~
auto &v = scratch.get_values("name", sol, dummy);
~~~
which is not possible if you don't provide the additional dummy argument. I find the second much easier to read.

Reply:
> As an aside, I think this interface is a bit clunky. What do you think about not returning a vector from this function? Then one would have to call extract_... once in the beginning, which would do the extraction. Every use of the extracted data would then require a call to get_.... This way, you'd call the same function every time, rather than calling extract_... the first time and get_... later on.

This was the original design. I'll revert back to this behaviour.

Reply:
> The current default of the template argument is double. Why not use typename VectorType::value_type?

The reason is that this is the only one function where this could be possible. All other functions do not have access to `VectorType::value_type`. My preference was to make the default uniform on all methods.

Reply:
Can't you just do
```
  template <typename Vector, typename Number=double>
  void do_something(Vector &v);
```
and the compiler will automatically use `Number==double`? I.e., in essence do the same as you do without ever making `Number` show up in the argument list? I thought that you could then just call `object.do_something(v)` and it will take `Number=double`, and if you want something different, you'd just write `object.template do_something<Number>()` -- which is admittedly clunkier, but will probably be rarely used, right?

Reply:
I don't think I understood your reasoning yet.